Remove explicit `IO` from our SMT interface

VictorCMiraldo opened the pull request:
# What

This PR does two large, but intertwined, pieces of work:
1. Isolates the interface to the SMT solver from the rest of the code, preparing `pirouette` to interact with a solver through:
    ```haskell
    solve :: SharedContext lang -> Problem lang res -> res
    ```
2. In this process, I also used the more modern `typed-process` to launch the SMT process instead of how `SimpleSMT` used to do it.
3. Cleans and refactors a lot of the symbolic evaluation engine

Partially applying something to `solve` should cause further calls to share that shared context. Here's a gist showing a little experiment I did in this direction: https://gist.github.com/VictorCMiraldo/b7c448bb2bcf4b49962082970d61f64a

It doesn't use that `solve` function just yet, we need some further cleanup (#106) to start using it. Nevertheless, the interface with the symbolic evaluation engine is already defined in terms of a [_problem_ GADT](https://github.com/tweag/pirouette/pull/99/files#diff-50e4bfb2e5aeae52f3faf04f67fb500e6765a837342513c57e9ea2ca9757062bR61):
```haskell
data SolverProblem lang :: * -> * where
  CheckProperty :: (LanguagePretty lang) => CheckPropertyProblem lang -> SolverProblem lang PruneResult
  CheckPath :: CheckPathProblem lang -> SolverProblem lang Bool
```
It is only inside the [interpretation](https://github.com/tweag/pirouette/pull/99/files#diff-50e4bfb2e5aeae52f3faf04f67fb500e6765a837342513c57e9ea2ca9757062bR119) of those problems that we interact directly with the solver; anywhere else in the symbolic engine we just call `solveProblem` for now and in a next PR, will become just `solve`.

# Why

Having an explicit `IO` monad (or a `MonadIO m` constraint) simply because we need to talk to an SMT solver is a bad idea. For one, it means we cannot be lazy enough and are always performing more work than we should (#97); Secondly, this implies an unnecessary sentimentality to our backend: why can't we easily explore disjoint branches in parallel?

VictorCMiraldo commented:
The last commit broke our tests. Our session with the solver now only sends empty `assert` statements:

```
[send: 120238] (set-option :print-success true )
[recv: 120238] success
[send: 120238] (set-option :produce-models true )
[recv: 120238] success
[send: 120238] (set-logic ALL )
[recv: 120238] success
[send: 120238] (declare-datatype pir_Delta ((pir_D (pir_D_1 Int ) (pir_D_2 Int ) ) ) )
[recv: 120238] success
[send: 120238] (declare-fun pir_odd (Int ) Bool )
[recv: 120238] success
[send: 120238] (declare-fun pir_even (Int ) Bool )
[recv: 120238] success
[send: 120238] (declare-fun pir_and (Bool Bool ) Bool )
[recv: 120238] success
[send: 120238] (declare-fun pir_snd (pir_Delta ) Int )
[recv: 120238] success
[send: 120238] (declare-fun pir_fst (pir_Delta ) Int )
[recv: 120238] success
[send: 120238] (declare-fun pir_ohearn (pir_Delta ) pir_Delta )
[recv: 120238] success
[send: 120238] (push 1 )
[recv: 120238] success
[send: 120238] (declare-fun pir___result () pir_Delta )
[recv: 120238] success
[send: 120238] (declare-fun pir_x () pir_Delta )
[recv: 120238] success
[send: 120238] (check-sat )
[recv: 120238] sat
[send: 120238] (pop 1 )
[recv: 120238] success
[send: 120238] (push 1 )
[recv: 120238] success
[send: 120238] (declare-fun pir___result () pir_Delta )
[recv: 120238] success
[send: 120238] (declare-fun pir_s0 () Int )
[recv: 120238] success
[send: 120238] (declare-fun pir_s1 () Int )
[recv: 120238] success
[send: 120238] (declare-fun pir_s2 () Int )
[recv: 120238] success
[send: 120238] (declare-fun pir_s3 () Int )
[recv: 120238] success
[send: 120238] (declare-fun pir_x () pir_Delta )
[recv: 120238] success
[send: 120238] (assert true ) -- HERE: empty assert!!
[recv: 120238] success
[send: 120238] (assert true )
[recv: 120238] success
[send: 120238] (check-sat )
[recv: 120238] sat
```

For the record, this is the same part of the session with the solver from commit 96b78f4a6a9c9ff1270319265531544c48a17979 (with the compilation error fixed)

```
[send: 115976] (set-option :print-success true )
[recv: 115976] success
[send: 115976] (set-option :produce-models true )
[send: 115976] (set-logic ALL )
[recv: 115976] success
[send: 115976] (declare-datatype pir_Delta ((pir_D (pir_D_1 Int ) (pir_D_2 Int ) ) ) )
[recv: 115976] success
[send: 115976] (declare-fun pir_odd (Int ) Bool )
[recv: 115976] success
[send: 115976] (declare-fun pir_even (Int ) Bool )
[recv: 115976] success
[send: 115976] (declare-fun pir_and (Bool Bool ) Bool )
[recv: 115976] success
[send: 115976] (declare-fun pir_snd (pir_Delta ) Int )
[recv: 115976] success
[send: 115976] (declare-fun pir_fst (pir_Delta ) Int )
[recv: 115976] success
[send: 115976] (declare-fun pir_ohearn (pir_Delta ) pir_Delta )
[recv: 115976] success
[send: 115976] (push 1 )
[recv: 115976] success
[send: 115976] (declare-fun pir___result () pir_Delta )
[recv: 115976] success
[send: 115976] (declare-fun pir_x () pir_Delta )
[recv: 115976] success
[send: 115976] (check-sat )
[recv: 115976] sat
[send: 115976] (pop 1 )
[recv: 115976] success
[send: 115976] (push 1 )
[recv: 115976] success
[send: 115976] (declare-fun pir___result () pir_Delta )
[recv: 115976] success
[send: 115976] (declare-fun pir_s0 () Int )
[recv: 115976] success
[send: 115976] (declare-fun pir_s1 () Int )
[recv: 115976] success
[send: 115976] (declare-fun pir_s2 () Int )
[recv: 115976] success
[send: 115976] (declare-fun pir_s3 () Int )
[recv: 115976] success
[send: 115976] (declare-fun pir_x () pir_Delta )
[recv: 115976] success
[send: 115976] (assert (= pir_x ((as pir_D pir_Delta ) pir_s2 pir_s3 ) ) ) -- HERE: Non-empty assert
[recv: 115976] success
[send: 115976] (assert (= pir___result ((as pir_D pir_Delta ) pir_s0 pir_s1 ) ) )
[recv: 115976] success
[send: 115976] (check-sat )
[recv: 115976] sat
```

VictorCMiraldo commented:
While stripping everything Pirouette-specific from our API with the SMT solver, I noticed that we rely on `PirouetteReadDefs` in multiple places, which is perfectly fine. The problem is that `PirouetteReadDefs` has a few too many dependencies, in particular, the Logger monad and `PrtOpts` are no longer used, pirouette is not meant as a standalone executable anymore. These should be removed at some point.

0xd34df00d reviewed (approved):
All looks good, modulo a couple of nitpicks and maybe `pushMStack`/`popMStack` stuff.

VictorCMiraldo commented on src/Pirouette/SMT/Constraints.hs:
```diff
@@ -60,14 +61,15 @@ data Branch lang meta = Branch
     newTerm :: TermMeta lang meta
   }
 
+-- TODO: Maybe this should be merged with 'LanguageSMT'
```
Our implementation requires both instances to run the symbolic evaluation, so I guess it doesn't really matter, any `lang` must be `LanguageSMT lang` and `LanguageSMTBranches lang`.

I'm under the impression that these classes were separated because certain types used inside `LanguageSMTBranches` are not available where `LanguageSMT` is at and would require some puzzle solving to find non-cyclical imports to satisfy everything. I decided to leave it untouched for now.

serras replied:
This might be a good moment to rename those classes. The general idea is:

```
LanguageBuiltins --> defines the built-ins (both terms and types)
  |     \
  |      LanguageBuiltinTypes --> defines the typing rules of each built-in term
  |
LanguageSMT --> defines translation of built-ins into SMT
  |             (not every term can be translated)
  |
LanguageSMTBranches --> defines at which points the symbolic evaluator has to do sth. special
```

on that sense, `LanguageSMTBranches` could well be renamed to `LanguageSymEval` or something similar.

To give you a difference, the `LanguageSMT` translation of `if ... then ... else ...` uses the `ifthenelse` primitive in Z3. However, the `LanguageSMTBranches` translation specifies that when an `if` is found, two *new* branches must be created:
1. Add `condition = True` to the list of known things, continue with the `then` term,
2. Add `condition = False` to the list of known things, continue with the `else` term.

VictorCMiraldo replied:
I won't do that in this PR, I'll do it on the next, which will remove the `Logger` too. I like `LanguageSymEval`!

VictorCMiraldo commented on src/Pirouette/SMT/Monadic.hs:
```diff
@@ -102,38 +103,50 @@ checkSat = do
+
+-- |If a function can be declared as an uninterpreted function, returns
+-- the type of its arguments and the type of its result.
+supportedUninterpretedFunction ::
```
previously, we were using `declareAsManyUninterpretedFunctinosAsPossible` inside `initSolver`, and `initSolver` would return the list of names that managed to be declared. Those that we don't "support", just get silently ignored, which was fine but it adds a dependency between `initSolver` and everything else. 

I split `declareAsManyUninterpretedFunctionsAsPossible` into two: first give me the list of functions that I know I can declared (because we've managed to translate their types already), then declare then in `initSolver`. Now, I don't need the result of `initSolver` to have the set of `knownNames`.

I'll remove the old `declareAsManyUninterpretedFunctionsAsPossible` and will update the comment.

VictorCMiraldo replied:
(I'll keep this unresolved so the comment is always visible, breaking the data dependencies is an important point of this PR)

VictorCMiraldo commented on src/Pirouette/SMT/Monadic.hs:
```diff
@@ -102,38 +103,50 @@ checkSat = do
+  let (args, result) = R.tyFunArgs funTy
+  (args', _) <- runWriterT $ mapM translateType args
+  (result', _) <- runWriterT $ translateType result
+  return (args', result')
```
`declareUninterpretedFunction` was deleted because it was unused. Better than merging it! :)

VictorCMiraldo commented on src/Pirouette/SMT/Monadic.hs:
```diff
@@ -161,17 +174,12 @@ declareUninterpretedFunctions decls orderedNames = do
+declareAsManyUninterpretedFunctionsAsPossible ds = do
+  forMaybeM ds $ \(fnName, fnDef) ->
+    (Just <$> declareUninterpretedFunction fnName fnDef)
+      `catchError` (\_ -> return Nothing)
```
You're correct, but that was deleted as it was unused.

VictorCMiraldo commented on src/PureSMT.hs:
```diff
@@ -0,0 +1,76 @@
+import System.IO.Unsafe (unsafePerformIO)
+
+-- The api to the SMT solver is simple; all the user has to do is inform us how to (A) initialize
+-- the solver with a shared context and (B) solve different problems. The @problems@ is supposed
+-- to be a GADT in order to fix the different result type of solving different problems.
```
Good catch, thanks! I need to update almost all comments on that module!

VictorCMiraldo commented on src/PureSMT.hs:
```diff
@@ -0,0 +1,76 @@
+  -- we end up with a list of MVars, which we will protect in another MVar.
+  allProcs <- launchAll @domain ctx >>= newMStack
+
+  -- Finally, we return the actual length function, which is ran in a 'catch'
```
:laughing:

serras commented on src/PureSMT.hs:
```diff
@@ -0,0 +1,76 @@
+newMStack xs = newMVar xs
+
+pushMStack :: a -> MStack a -> IO ()
+pushMStack a q = do
+  mas <- tryTakeMVar q
+  case mas of
+    Nothing -> putMVar q [a]
+    Just as0 -> putMVar q (a:as0)
+
+popMStack :: MStack a -> IO a
+popMStack q = do
+  xss <- takeMVar q
+  case xss of
+    [] -> error "invariant disrespected; MStack should never be empty"
+    [x] -> return x
+    (x:xs) -> putMVar q xs >> return x
```
I think we should definitely use [`TMVar`](https://hackage.haskell.org/package/stm-2.5.0.2/docs/Control-Concurrent-STM-TMVar.html) instead, which is guaranteed to never have race conditions.

VictorCMiraldo replied:
We can't use `modifyMVar` because we have to leave the `MVar` empty in case we'd put an empty list there. This is what guarantees that other threads are just waiting for a worker to be available at their respective `popMStack`

VictorCMiraldo replied:
But then we have to bring in `STM`; I modified `MStack a` to be defined as `(MVar (), MVar [a])`, whereas the first `MVar` works as a lock. That should be just a poor man `TMVar`

VictorCMiraldo commented on src/PureSMT.hs:
```diff
@@ -0,0 +1,76 @@
+    ms <- popMStack allProcs
+    r <- withMVar ms $ \solver -> do
+      solveProblem @domain problem solver
+    pushMStack ms allProcs
```
For now, we do nothing and `pushMStack` will be executed; that's undesirable indeed. I left a comment and will create an issue for us to think this through in a separate instance.

serras commented:
> While stripping everything Pirouette-specific from our API with the SMT solver, I noticed that we rely on `PirouetteReadDefs` in multiple places, which is perfectly fine. The problem is that `PirouetteReadDefs` has a few too many dependencies, in particular, the Logger monad and `PrtOpts` are no longer used, pirouette is not meant as a standalone executable anymore. These should be removed at some point.

This would be very welcome! I've struggled with this in the past, mostly because of `PirouetteReadDefs` being defined as its own `Reader` thing, making it difficult to compose with others.

VictorCMiraldo commented:
@0xd34df00d thanks for the review! I fixed all I could on this iteration and I added a lock to `MStack` to make sure no race condition can happen. That is untested code and I'll resort to really using it in the next PR.

VictorCMiraldo commented:
> This would be very welcome! I've struggled with this in the past, mostly because of PirouetteReadDefs being defined as its own Reader thing, making it difficult to compose with others.

@serras, I hope to do this today!
